src/tree/gpu_hist/evaluate_splits.cu

```diff
@@ -8,316 +8,269 @@
-
-    __shared__ cub::KeyValuePair<int, float> block_max;
-    if (threadIdx.x == 0) {
-      block_max = best;
-    }
-
-    cub::CTA_SYNC();
-
-    // Best thread updates split
-    if (threadIdx.x == block_max.key) {
-      update_fn(missing_left, scan_begin, gain, fidx, missing, bin, inputs,
-                best_split);
-    }
-    cub::CTA_SYNC();
-  }
-}
-
-template <int BLOCK_THREADS, typename GradientSumT>
-__global__ void EvaluateSplitsKernel(
     EvaluateSplitInputs<GradientSumT> left,
-    EvaluateSplitInputs<GradientSumT> right,
-    TreeEvaluator::SplitEvaluator<GPUTrainingParam> evaluator,
-    common::Span<DeviceSplitCandidate> out_candidates) {
-  // KeyValuePair here used as threadIdx.x -> gain_value
-  using ArgMaxT = cub::KeyValuePair<int, float>;
-  using BlockScanT =
-      cub::BlockScan<GradientSumT, BLOCK_THREADS, cub::BLOCK_SCAN_WARP_SCANS>;
-  using MaxReduceT = cub::BlockReduce<ArgMaxT, BLOCK_THREADS>;
-
-  using SumReduceT = cub::BlockReduce<GradientSumT, BLOCK_THREADS>;
-
-  union TempStorage {
-    typename BlockScanT::TempStorage scan;
-    typename MaxReduceT::TempStorage max_reduce;
-    typename SumReduceT::TempStorage sum_reduce;
-  };
-
-  // Aligned && shared storage for best_split
-  __shared__ cub::Uninitialized<DeviceSplitCandidate> uninitialized_split;
-  DeviceSplitCandidate& best_split = uninitialized_split.Alias();
-  __shared__ TempStorage temp_storage;
-
-  if (threadIdx.x == 0) {
-    best_split = DeviceSplitCandidate();
+    EvaluateSplitInputs<GradientSumT> right) {
+  auto l_n_features = left.feature_segments.empty() ? 0 : left.feature_segments.size() - 1;
+  auto r_n_features = right.feature_segments.empty() ? 0 : right.feature_segments.size() - 1;
+  if (!(r_n_features == 0 || l_n_features == r_n_features)) {
+    throw std::runtime_error("Invariant violated");
   }
 
-  __syncthreads();
-
-  // If this block is working on the left or right node
-  bool is_left = blockIdx.x < left.feature_set.size();
-  EvaluateSplitInputs<GradientSumT>& inputs = is_left ? left : right;
+  std::size_t size = left.gradient_histogram.size() + right.gradient_histogram.size();
+  CHECK_LE(size, static_cast<std::size_t>(std::numeric_limits<uint32_t>::max()));
+  auto count_iter = dh::MakeTransformIterator<thrust::tuple<uint32_t, bool>>(
+      thrust::make_counting_iterator<uint32_t>(0),
+      [size = static_cast<uint32_t>(size)] __device__(uint32_t i) {
+        // Generate sequence of length (size * 2):
+        // 0 1 2 3 ... (size-2) (size-1) (size-1) (size-2) ... 2 1 0
+        // At each element, attach a boolean indicating whether the sequence is going forward or
+        // backward.
+        if (i < size) {
+          return thrust::make_tuple(i, true);
+        } else if (i < size * 2) {
+          // size <= size < size * 2
+          return thrust::make_tuple(size * 2 - 1 - i, false);
+        } else {
+          return thrust::make_tuple(static_cast<uint32_t>(0), false);
```

Comment:
Does this happen?

Reply:
It should not, I put it there just in case.

Reply:
You can put a `SPAN_CHECK` here as runtime assert.